### ISSUE TITLE
Mod Support: Add support for server-provided assets

### ISSUE THREAD

**Issue by Kaffeine** [ddnet/ddnet#7296]
**The purpose of the issue to discuss and track ideas toward server-provided assets.**

It would be cool to have support for:
- [server-side skins](https://github.com/ddnet/ddnet/issues/7055#issuecomment-1686266191)
- [server-side weapons, etc](https://github.com/ddnet/ddnet/issues/7042#issuecomment-1746827004) (@VoxelDoesCode)
- [server-side mapres](https://github.com/ddnet/ddnet/issues/5165)
- etc

Probably we can invent a generic protocol allowing clients to request missing resources from the server (or maybe some side server).

**Comment by heinrich5991** [comment]
Another important question to solve: How does this interact with demos?

**Comment by Kaffeine** [comment]
For the first step it can be the same as for custom skins (skin name saved and used during playback if found) and/or custom gameskin (totally ignored). (Downloaded) map resources are probably embedded into the map (the map is reconstructed on the client side, just like [you said](https://github.com/ddnet/ddnet/issues/5165#issuecomment-1137862467)).

In this issue I'd rather focus on the generic protocol for the assets delivery. I don't understand how to do this part properly on the network level and I need your guidance here. After that I can experiment with different approaches and we'll pick the most suitable one.

I think experiments would quickly show weak and missing points of the idea but personally I'm yet to get there. I dunno how to reliably transfer a big data chunk using tw protocol and keep client connection responsible (in terms of gameplay). Maybe it is enough to send all data together or after the map downloaded and before the client `m_State == CClient::STATE_READY`.

**Comment by heinrich5991** [comment]
I don't like the idea of making demos non-self-contained, but I agree that focusing on the protocol first seems worthwhile.

> I think experiments would quickly show weak and missing points of the idea but personally I'm yet to get there. I dunno how to reliably transfer a big data chunk using tw protocol and keep client connection responsible (in terms of gameplay).

The Teeworlds protocol doesn't really have a nice way of doing that. We're basically hand-rolling a new way each time we need to do this. The console has issues with responsiveness/connection drops because it's not throttled, votes are sent in a throttled fashion (but have no end of list indicator), map data is sent in chunks, …

Maybe it's a good idea to only serve a https link and let the client download the data from there? This would obviously make the server not self-contained which is bad because we might run into servers that you can't join because the HTTPS download is misconfigured. QUIC support (#6961) would help here, but it being finished is not something we can count on.

Sooo… hand out a https link and hand-roll a fallback protocol similar to the map download?

**Comment by heinrich5991** [comment]
> It would be cool to have support for:
> 
>     * [server-side skins](https://github.com/ddnet/ddnet/issues/7297)
> 
>     * [server-side weapons, etc](https://github.com/ddnet/ddnet/issues/7042#issuecomment-1746827004) (@VoxelDoesCode)
> 
>     * [server-side mapres](https://github.com/ddnet/ddnet/issues/5165)
> 
>     * [server-side sounds](https://github.com/ddnet/ddnet/issues/7299)

Hmm… Maybe all of the above could simply be put into the map. That would solve all the issues(?).

**Comment by VoxelDoesCode** [comment]
Wait, if they can be put into the map, would this enable ddnet levels to customize the skins/weapons/etc.?

**Comment by heinrich5991** [comment]
Note: This is talking about a feature that isn't implemented.

> Wait, if they can be put into the map, would this enable ddnet levels to customize the skins/weapons/etc.?

This would allow mods to customize skins/weapons/etc. This is separate, but required, for allowing DDNet maps to change skins/weapons/etc.

**Comment by Kaffeine** [comment]
> Maybe all of the above could simply be put into the map. That would solve all the issues(?).

Yes, it would solve virtually all the issues of the data fetching. Sounds good for mod support but won't help https://github.com/ddnet/ddnet/issues/5165 at all.

In infclass we already have players complaining about long map downloads — map rotation happens every 10-15 minutes, some maps already have big resources. 10 seconds warmup is not enough for all players.

**Comment by heinrich5991** [comment]
Do you already employ HTTPS map download? The nice thing about putting the resources into the map is that it decouples #5165 from this issue completely.

**Comment by M0REKZ** [comment]
Since #9822 was closed as duplicate i'll rewrite the idea here:

Would be good to also have Server-side entities for the reason i mentioned in that issue, maybe should be done by adding a flag for layers so they are rendered on entities mode even if they have alpha on 0

However there are some things:

* If the flag is set in Game, Front, Switch, Tele, Tune, or Speedup layers, should it override the entities shown for those?
* Should it work only for tile layers or also include quads? (Useful for KOG maps and https://github.com/ddnet/ddnet/issues/6683#issuecomment-1566949792)
* Should it support CTiles only? Or tiles with any size? Maybe by just reading every tile as CTile and omitting the extra data?

**Comment by Bamcane** [comment]
I had written a long text, but then I found what I said isn't what you wanted.

May you can try the solution of [InfClassR](https://github.com/infclass/teeworlds-infclassR].

And then add the method I originally wanted to say:
I guess we can add a net message like CNetMsg_SvSoundGlobal did.
It will force the client to use one of the images in the map as entities image.
Also, adding a net message is good for some multi-modes mod.

This could also help with some mode that don't want players to use their own entities layer image.
such as [HiddenWorld](https://github.com/landmineHQ/ddnet-hidden), [JailHide](https://github.com/bamcane/teeworlds-hide), and some fng modes.

**Comment by M0REKZ** [comment]
Idk what you mean with the infclass solution, as I remember it also has some custom entities layers but they are only visible on their custom client.

However the idea of making a new net message sounds cool, maybe it could include info about the CTile size and the m_Index position (I think Tile size was saved in the map too but I'm not sure)

**Comment by Bamcane** [comment]
It would make the map bigger if we put all the things into it.
Another way you could find in https://github.com/TeeworldsArchive/teeworlds/pull/46
As custom resource is not a part of engine, it should be new game messages.
Currently, I only implement playing custom sound(.opus) and using custom image(.png) as entity texture.
The download process has been finished, but need to discuss if we should pack it as our teeworlds datafile instead of sending raw data.
We could also implement #5165 based on this (maybe?)

**Comment by Bamcane** [comment]
I think another issue is we pack the uuid into it, maybe we could use a int as hash or index instead, but that would depends on how server implement.

**Comment by heinrich5991** [comment]
> It would make the map bigger if we put all the things into it.

Yes. The point of that is that demos contain all the data they need for playback.

> Another way you could find in [TeeworldsArchive/teeworlds#46](https://github.com/TeeworldsArchive/teeworlds/pull/46)
> As custom resource is not a part of engine, it should be new game messages.
> Currently, I only implement playing custom sound(.opus) and using custom image(.png) as entity texture.

I think we shouldn't bolt-on different download mechanisms for different parts of the map. Your suggestion seems to transport the map images with a protocol different from the one of the on-demand stuff. Instead, we should design a map format so that maps can be partially downloaded, like in #5165.

**Comment by Bamcane** [comment]
> I think we shouldn't bolt-on different download mechanisms for different parts of the map. Your suggestion seems to transport the map images with a protocol different from the one of the on-demand stuff. Instead, we should design a map format so that maps can be partially downloaded, like in [#5165](https://github.com/ddnet/ddnet/issues/5165).

Make map be partially downloaded is another issue. What this issue wanted is only for mods.
If everything is a part of map though it's not in the map file, that would be terrible either.

**Comment by heinrich5991** [comment]
> What this issue wanted is only for mods.

Adding assets to maps would work for mods, no?

**Comment by fokkonaut** [comment]
What about https://github.com/teeworlds/teeworlds/compare/master...LordSk:duck?
@LordSk did some huge work and showed it to me a few years ago. This should not go unnoticed or wasted!
